File: exercises/ddd_application_layer/java/src/test/java/application/FillOutBasicDetailsServiceTest.java
package application;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import org.junit.jupiter.api.Test;

import application.FillOutBasicDetailsService.FillOutBasicDetailsCommand;
import domain.application.Application;
import helpers.RecordingApplicationRepository;

public class FillOutBasicDetailsServiceTest {

    @Test
    public void testFillOutBasicDetails() {
        RecordingApplicationRepository repo = new RecordingApplicationRepository();
        FillOutBasicDetailsService service = new FillOutBasicDetailsService(repo);

        service.execute(new FillOutBasicDetailsCommand(
            "Taylor Applicant",
            "441234567890",
            "taylor@example.com"
        ));

        assertEquals(1, repo.saved.size());
        Application saved = repo.saved.get(0);
        assertEquals("Taylor Applicant", saved.getFullName().getValue());
        assertEquals("441234567890", saved.getPhoneNumber().getValue());
        assertEquals("taylor@example.com", saved.getEmailAddress().getValue());
    }

    @Test
    public void testFillOutBasicDetailsInvalidEmail() {
        RecordingApplicationRepository repo = new RecordingApplicationRepository();
        FillOutBasicDetailsService service = new FillOutBasicDetailsService(repo);

        assertThrows(RuntimeException.class, () -> {
            service.execute(new FillOutBasicDetailsCommand(
                "Taylor Applicant",
                "441234567890",
                "not-an-email"
            ));
        });

        assertEquals(0, repo.saved.size());
    }
}

File: exercises/ddd_infrastructure_layer/typescript/test/application/services/FillOutBasicDetailsService.test.ts
import { describe, expect, it } from "bun:test";
import { FillOutBasicDetailsService } from "../../../src/application/services/FillOutBasicDetailsService";
import { RecordingApplicationRepository } from "../../helpers/RecordingApplicationRepository";

describe("FillOutBasicDetailsService", () => {
  it("persists the new application and publishes resulting events", async () => {
    const repo = new RecordingApplicationRepository();
    const service = new FillOutBasicDetailsService(repo);

    const result = await service.execute({
      fullName: "Taylor Applicant",
      phoneNumber: "441234567890",
      emailAddress: "taylor@example.com",
    });

    expect(result.applicationId).toBeTruthy();

    expect(repo.saved.length).toBe(1);
    const [saved] = repo.saved;
    expect(saved!.fullName!.value).toBe("Taylor Applicant");
    expect(saved!.phoneNumber!.value).toBe("441234567890");
    expect(saved!.emailAddress!.value).toBe("taylor@example.com");
  });

  it("wraps domain validation errors", async () => {
    const repo = new RecordingApplicationRepository();
    const service = new FillOutBasicDetailsService(repo);

    expect(
      service.execute({
        fullName: "Taylor Applicant",
        phoneNumber: "441234567890",
        emailAddress: "not-an-email",
      })
    ).rejects.toThrow();

    expect(repo.saved.length).toBe(0);
  });
});
